src/exprs.c

```diff
@@ -811,165 +811,117 @@ static Obj EvalPermExpr(Expr expr)
+**  'EvalListTildeExpr' evaluates the     list  expression, i.e., not     yet
+**  evaluated list, <expr> to a list value.  The difference to 'EvalListExpr'
+**  is that  in <expr> there are  occurrences of '~'  referring to  this list
+**  value.
 **
-**  When 'tildeInUse' is 1, use a slower code path which ensures the list is
-**  in a valid state after each element is added and handles the elements
-**  and TNUM of the list changing whenever a new element is constructed.
+**  Note that we do not track here whether the list is dense, as this can be
+**  changed by code involving a tilde expression, as in this example:
+**      x := [1,,3,function(x) x[2]:=2; return 4; end(~)];
 **
+**  For similar reasons, we must deal with the possibility that the list we
+**  are creating changes its representation, and thus must use ASS_LIST
+**  instead of SET_ELM_PLIST.
 */
-static ALWAYS_INLINE Obj ListExpr1(Expr expr, Int tildeInUse)
+static Obj EvalListTildeExpr(Expr expr)
 {
     Obj                 list;           /* list value, result              */
+    Obj                 tilde;          /* old value of tilde              */
+    Obj                 sub;            /* value of a subexpression        */
     Int                 len;            /* logical length of the list      */
+    Int                 i;              /* loop variable                   */
 
-    /* get the length of the list                                          */
+    // get the length of the list
     len = SIZE_EXPR(expr) / sizeof(Expr);
 
-    /* allocate the list value                                             */
-    if ( 0 == len ) {
-        list = NEW_PLIST( T_PLIST_EMPTY, len );
-    }
-    else {
-        list = NEW_PLIST( T_PLIST, len );
-    }
-
-    SET_LEN_PLIST(list, tildeInUse ? 0 : len);
+    // list expressions with tilde cannot be empty
+    GAP_ASSERT(len > 0);
 
-    /* return the list                                                     */
-    return list;
-}
+    // allocate the list value
+    list = NEW_PLIST(T_PLIST, len);
 
-static ALWAYS_INLINE void ListExpr2(Obj list, Expr expr, Int tildeInUse)
-{
-    Obj                 sub;            /* value of a subexpression        */
-    Int                 len;            /* logical length of the list      */
-    Int                 i;              /* loop variable                   */
-    Int                 posshole;       /* initially 0, set to 1 at
-                                           first empty position, then
-                                           next full position causes
-                                           the list to be made
-                                           non-dense */
+    // remember the old value of '~'
+    tilde = STATE(Tilde);
 
-    /* get the length of the list                                          */
-    len = SIZE_EXPR(expr) / sizeof(Expr);
+    // assign the list to '~'
+    STATE(Tilde) = list;
```

Comment:
Huh... I just thought about all this, and now I have serious doubts whether this PR is correct -- or rather, whether the code it refactors was correct; I now think we must *always* use the "slow" path; or else we need a third expression type T_LIST_NESTED_IN_TILDE_EXPR or so (and all these changes are also needed for records).

Reply:
OK, when I wrote the above, I thought the following might happen: while inside a `T_LIST_EXPR` which is nested inside a `T_TILDE_LIST_EXPR`, we can use `~` to access subobjects of the parent list. I thought we could use that to modify the nested list as it was being constructed. However, this is not true, because we only store a reference to the nested list into the parent list after the nested list has been fully constructed; any access to it after that, however, is fine.

So, false alarm. Phew :-)